cloud topics: introduce cluster::epoch_generator

rockwotj opened the pull request:
This creates a cluster services instance for cloud topics that is based on the controller committed offset. Additionally there is a fiber that ensures that periodically we increment the last committed offset (currently hard coded to 1min) if there are no other updates to the controller log.


- **raft: add committed offset to mux_state_machine**
- **cluster: introduce cluster epoch generator**
- **cluster: add epoch generator to the controller**
- **cloud_topics: plumb in the real cluster services**

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

rockwotj committed:
cluster: introduce cluster epoch generator

This can be used to implement the cluster epoch service in cloud topics.

rockwotj committed:
cluster: add epoch generator to the controller

rockwotj committed:
cloud_topics: plumb in the real cluster services

rockwotj committed:
application: wait to start cloud topics subsystem until after controller

We need to wait to start cloud topics until after the controller comes
online, otherwise we'll try to access other parts of the system that are
not ready yet. This pattern of contruct in wire_up, and start in
start_runtime_services matches what other subsystems do. I suppose cloud
topics was self contained before this change and it could be started
whenever. As we add more functionality (like RPCs, etc), we may need to
break up the cloud topics boot sequence.

rockwotj commented on src/v/cluster/cluster_epoch_generator.cc:
```diff
@@ -0,0 +1,130 @@
+            last_offset = new_offset;
+            continue;
+        }
+        auto result = co_await stm.quorum_write_empty_batch(
+          ss::lowres_clock::now() + 3s);
```
In my local setup this causes a bunch of extra snapshots to be taken - should we make the snapshot mechanism ignore these empty batch writes?

bharathv replied:
perhaps we should.. controller snapshots are huge (multi MB in some cases) and given the defaults there would be a lot of (unnecessary) snapshotting IMO.

rockwotj replied:
I'll add a JIRA for it. Do we always eagerly create a snapshot? I feel like they should only be every N bytes or N commits to the log, it seems better to replay a couple of updates than rewrite multi-MiB for every update (especially since large scale clusters/snapshots are more likely to get more udpates)

bharathv replied:
its not for every update.. we do batch data but on a time basis (`controller_snapshot_max_age_sec`).. all the commits in this window are snapshotted together.. the default is a bit aggressive tho

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69467-->
<!--a4111a26-dccf-487e-9d8f-ee27b61ee48e-->
<!--a2824775-6be8-455d-bba5-a1a556f86c61-->
<!--title-->
## Retry command for [Build#69467](https://buildkite.com/redpanda/redpanda/builds/69467)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTest.test_write
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTxTest.test_write
```
<!--/body-->

rockwotj commented:
Test failures are due to getting an epoch while bootstrapping. I will change this to default to epoch 0, but would love to know if restarts are prone to this negative committed offset behavior as well.

rockwotj committed:
raft: update docs for raft::state_machine::bootstrap_committed_offset

My reading of the current comment made me think this was only returning
the committed offset at bootstrap time (that is was fixed for the
lifetime of the object), which isn't true.

rockwotj commented:
Force push: fix issue getting committed offset when the STM is bootstrapping still

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69480-->
<!--0a0c8acb-75f2-4875-9742-0268d7a47fbf-->
<!--fbdfcd9d-c990-4772-887c-570158e27405-->
<!--title-->
## Retry command for [Build#69480](https://buildkite.com/redpanda/redpanda/builds/69480)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/scram_test.py::ScramBootstrapUserTest.test_invalid_scram_mechanism@{"expect_fail":true,"mechanism":"sCrAm-ShA-512"}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--05b41922-5904-4a82-bf17-c2d3ea6a75b5-->
<!--TEST RESULTS-->
<!--cdc19e02-b42e-47bc-a4bf-931d58936118-->
<!--TEST RESULTS-->
<!--da6c0ac2-fcdb-492b-af9f-9393acc6e0c0-->
<!--TEST RESULTS-->
<!--67783711-8a20-48a5-81c8-f02a996ce1e4-->
<!--TEST RESULTS-->
<!--17bab5ae-5948-425d-91f7-3cfbd21fad37-->
<!--TEST RESULTS-->
<!--661c2c55-13cc-4b92-89fc-2c20889999bf-->
<!--TEST RESULTS-->
<!--4b249d18-c8b6-455e-af0d-3032235a5a91-->
<!--TEST RESULTS-->
<!--90ef6acc-b9cd-4d33-a2f2-f19596d2d759-->
<!--TEST RESULTS-->
<!--32787be5-f0ca-47ff-a3b0-4646ce56a00a-->
<!--TEST RESULTS-->
<!--05d14d41-09e5-4db8-bced-074984e11e7b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69480">build#69480</a></summary>


| test_class             | test_method                   | test_arguments                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------|:------------------------------|:----------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| FeaturesMultiNodeTest  | test_license_upload_and_query |                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69480#019834c5-c31c-4a0e-8e0a-503b293b0a8e | FLAKY         | 16/21    | upstream reliability is '98.04878048780488'. current run reliability is '76.19047619047619'. drift is 21.8583 and the allowed drift is set to 50. The test should PASS |
| ScramBootstrapUserTest | test_invalid_scram_mechanism  | {"expect_fail": true, "mechanism": "sCrAm-ShA-512"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69480#019834c3-2e2e-4ca2-aaab-bdee09b0608f | FAIL          | 0/1      |                                                                                                                                                                        |
| ScramBootstrapUserTest | test_invalid_scram_mechanism  | {"expect_fail": true, "mechanism": "sCrAm-ShA-512"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69480#019834c5-c31c-478d-b1d9-0a3e19944c54 | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69511">build#69511</a></summary>


| test_class                                             | test_method                                            | test_arguments                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:-------------------------------------------------------|:-------------------------------------------------------|:----------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| ScramBootstrapUserTest                                 | test_invalid_scram_mechanism                           | {"expect_fail": true, "mechanism": "sCrAm-ShA-512"} | integration | https://buildkite.com/redpanda/redpanda/builds/69511#01983732-dd27-4321-b52e-5689870ac2bf | FAIL          | 0/1      |          |
| src/v/cluster_link/tests/task_manager_integration_test | src/v/cluster_link/tests/task_manager_integration_test |                                                     | unit        | https://buildkite.com/redpanda/redpanda/builds/69511#01983718-716a-4c35-a298-c0991b7378d1 | FAIL          | 0/1      |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69691">build#69691</a></summary>


| test_class                | test_method             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:--------------------------|:------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| ClusterEpochService       | FetchesLimitedToShard0  |                           | unit        | https://buildkite.com/redpanda/redpanda/builds/69691#01984379-6b2b-4e43-8950-8493e7194f89 | FAIL          | 0/1      |          |
| ClusterEpochService       | TestCaching             |                           | unit        | https://buildkite.com/redpanda/redpanda/builds/69691#01984379-6b2b-4e43-8950-8493e7194f89 | FAIL          | 0/1      |          |
| EndToEndCloudTopicsTest   | test_write              | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/69691#019843a3-cf18-4474-8ac1-21f887980541 | FAIL          | 0/1      |          |
| EndToEndCloudTopicsTxTest | test_write              | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/69691#019843a3-cf19-47b1-a6ed-ed688aad2c0a | FAIL          | 0/1      |          |
| CloudTopicsTest           | test_reconciler_uploads | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69691#019843a3-cf1a-4288-98f3-727650a8fe66 | FAIL          | 0/1      |          |
| CloudTopicsTest           | test_reconciler_uploads | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69691#019843a3-cf1b-47f7-b01e-ea5532e21421 | FAIL          | 0/1      |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69772">build#69772</a></summary>


| test_class                | test_method             | test_arguments                | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:--------------------------|:------------------------|:------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest   | test_write              | null                          | integration | https://buildkite.com/redpanda/redpanda/builds/69772#019851ed-7a01-4e4e-8bc0-705110606e2b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| EndToEndCloudTopicsTest   | test_write              | null                          | integration | https://buildkite.com/redpanda/redpanda/builds/69772#019851ef-941c-45e9-895c-17e558da9d09 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| EndToEndCloudTopicsTxTest | test_write              | null                          | integration | https://buildkite.com/redpanda/redpanda/builds/69772#019851ed-7a02-4393-8233-94424cbe59cc | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| EndToEndCloudTopicsTxTest | test_write              | null                          | integration | https://buildkite.com/redpanda/redpanda/builds/69772#019851ef-941d-4756-98fc-2adc7dee5abc | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| CloudTopicsTest           | test_reconciler_uploads | {"cloud_storage_type": 2}     | integration | https://buildkite.com/redpanda/redpanda/builds/69772#019851ed-7a03-412d-9736-23586da02948 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| CloudTopicsTest           | test_reconciler_uploads | {"cloud_storage_type": 2}     | integration | https://buildkite.com/redpanda/redpanda/builds/69772#019851ef-9411-4323-bcca-619763eebd02 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| CloudTopicsTest           | test_reconciler_uploads | {"cloud_storage_type": 1}     | integration | https://buildkite.com/redpanda/redpanda/builds/69772#019851ed-7a04-4164-861b-9ecbee94c6f7 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| CloudTopicsTest           | test_reconciler_uploads | {"cloud_storage_type": 1}     | integration | https://buildkite.com/redpanda/redpanda/builds/69772#019851ef-9412-4de4-be4b-2ea13ddae2e5 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| DataMigrationsApiTest     | test_transactions       | {"transfer_leadership": true} | integration | https://buildkite.com/redpanda/redpanda/builds/69772#019851ed-7a03-412d-9736-23586da02948 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69882">build#69882</a></summary>


| test_class         | test_method             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------|:------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudTopicsTest    | test_reconciler_uploads | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69882#01985843-3c8a-4b2d-97fa-fa92ca3bc4a0 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                |
| CloudTopicsTest    | test_reconciler_uploads | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69882#01985843-3c8c-49aa-89f7-f5ba2b65dcc0 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                |
| ShardPlacementTest | test_core_count_change  | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/69882#01985843-3c8e-4a5f-8905-54694c0a14e6 | FLAKY         | 19/21    | upstream reliability is '99.49238578680203'. current run reliability is '90.47619047619048'. drift is 9.0162 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69921">build#69921</a></summary>


| test_class      | test_method             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------|:------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudTopicsTest | test_reconciler_uploads | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69921#01985a0f-54d4-46e5-975a-fe394b3a808e | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| CloudTopicsTest | test_reconciler_uploads | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/69921#01985a11-dfb4-4b9b-8f85-e06a81787704 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| CloudTopicsTest | test_reconciler_uploads | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69921#01985a0f-54d5-4e9e-a7af-bf71d573e796 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| CloudTopicsTest | test_reconciler_uploads | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69921#01985a11-dfb4-4861-90cd-a510233b0f11 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| MaintenanceTest | test_maintenance_sticky | {"use_rpk": true}         | integration | https://buildkite.com/redpanda/redpanda/builds/69921#01985a0f-54d1-4be6-bb80-475f97642412 | FLAKY         | 20/21    | upstream reliability is '99.10112359550561'. current run reliability is '95.23809523809523'. drift is 3.86303 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69993">build#69993</a></summary>


| test_class          | test_method                   | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      |
|:--------------------|:------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterEpochService | IncrementMustHappenEventually |                           | unit        | https://buildkite.com/redpanda/redpanda/builds/69993#01985e65-ed00-48dc-bcf7-2a507f6f4c00 | FAIL          | 0/1      |                                                                                                                                                             |
| CloudTopicsTest     | test_reconciler_uploads       | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/69993#01985e94-efd8-48a9-bd06-661e95497c6e | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70020">build#70020</a></summary>


| test_class            | test_method                   | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------|:------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudTopicsTest       | test_reconciler_uploads       | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/70020#01986144-9ac2-474e-971e-ea96bc14239f | FLAKY         | 3/21     | upstream reliability is '100.0'. current run reliability is '14.285714285714285'. drift is 85.71429 and the allowed drift is set to 50. The test should FAIL            |
| CloudTopicsTest       | test_reconciler_uploads       | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/70020#0198614a-2673-4513-aa16-73d374999522 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL             |
| CloudTopicsTest       | test_reconciler_uploads       | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/70020#01986144-9ac3-4ab0-b8dd-ab6ee52af482 | FLAKY         | 4/21     | upstream reliability is '100.0'. current run reliability is '19.047619047619047'. drift is 80.95238 and the allowed drift is set to 50. The test should FAIL            |
| CloudTopicsTest       | test_reconciler_uploads       | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/70020#0198614a-2674-4716-b123-ce7664e251df | FLAKY         | 3/21     | upstream reliability is '100.0'. current run reliability is '14.285714285714285'. drift is 85.71429 and the allowed drift is set to 50. The test should FAIL            |
| FeaturesMultiNodeTest | test_license_upload_and_query | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/70020#0198614a-2671-4622-aef4-4a91c70f666f | FLAKY         | 20/21    | upstream reliability is '93.88586956521739'. current run reliability is '95.23809523809523'. drift is -1.35223 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70031">build#70031</a></summary>


| test_class                        | test_method                 | test_arguments             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------------|:----------------------------|:---------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudTopicsTest                   | test_reconciler_uploads     | {"cloud_storage_type": 2}  | integration | https://buildkite.com/redpanda/redpanda/builds/70031#019861f3-a8dc-4680-a22e-ed5ccbcf58f7 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| CloudTopicsTest                   | test_reconciler_uploads     | {"cloud_storage_type": 2}  | integration | https://buildkite.com/redpanda/redpanda/builds/70031#019861f9-8b0f-443a-8a28-ca88b8ba0643 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| CloudTopicsTest                   | test_reconciler_uploads     | {"cloud_storage_type": 1}  | integration | https://buildkite.com/redpanda/redpanda/builds/70031#019861f3-a8d0-4f16-8a9b-719129b4a0e3 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| CloudTopicsTest                   | test_reconciler_uploads     | {"cloud_storage_type": 1}  | integration | https://buildkite.com/redpanda/redpanda/builds/70031#019861f9-8b0f-4287-a8e3-c7a8b5835d66 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| VerifyConsumerOffsetsThruUpgrades | test_consumer_group_offsets | {"versions_to_upgrade": 3} | integration | https://buildkite.com/redpanda/redpanda/builds/70031#019861f9-8b0f-443a-8a28-ca88b8ba0643 | FLAKY         | 20/21    | upstream reliability is '99.73333333333333'. current run reliability is '95.23809523809523'. drift is 4.49524 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70050">build#70050</a></summary>


| test_class      | test_method             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                       |
|:----------------|:------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudTopicsTest | test_reconciler_uploads | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/70050#01986387-10da-470b-ab31-d03b840e5965 | FLAKY         | 12/21    | upstream reliability is '100.0'. current run reliability is '57.14285714285714'. drift is 42.85714 and the allowed drift is set to 50. The test should PASS  |
| CloudTopicsTest | test_reconciler_uploads | {"cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/70050#01986387-274e-4533-a115-1e5553ad21cd | FLAKY         | 8/21     | upstream reliability is '100.0'. current run reliability is '38.095238095238095'. drift is 61.90476 and the allowed drift is set to 50. The test should FAIL |
| CloudTopicsTest | test_reconciler_uploads | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/70050#01986387-10d4-4c73-864e-78f28dd5c413 | FLAKY         | 10/21    | upstream reliability is '100.0'. current run reliability is '47.61904761904761'. drift is 52.38095 and the allowed drift is set to 50. The test should FAIL  |
| CloudTopicsTest | test_reconciler_uploads | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/70050#01986387-274f-44d0-bff1-177d8520a9ac | FLAKY         | 6/21     | upstream reliability is '100.0'. current run reliability is '28.57142857142857'. drift is 71.42857 and the allowed drift is set to 50. The test should FAIL  |
</details>
<!--/body-->

bharathv commented on src/v/cluster/cluster_epoch_generator.cc:
```diff
@@ -0,0 +1,130 @@
+        if (!result) {
+            vlog(
+              clusterlog.warn,
+              "Failed to force increment epoch: {}",
```
isn't this chatty? returns success only on the controller leader?

rockwotj replied:
Ah yes thanks. I've ignored `raft::errc::not_leader` for now. However a better way might be to only have this loop run if the node is controller leader. Let me see how hard that is to plumb here.

bharathv commented on src/v/cluster/cluster_epoch_generator.cc:
```diff
@@ -0,0 +1,130 @@
+    }
+}
+
+ss::future<int64_t> cluster_epoch_generator::current_epoch() {
```
probably better to push the epoch to all shards (instead of pulling under a mutex) ?  we don't have a pile up of waiters..

I guess it really depends on whether we want to block the L0 writes if the epoch doesn't bump for some reason (loss of controller leadership or some weird edge cases). In the current design the epoch users would block but perhaps we should let them make progress?

rockwotj replied:
Goodpoint... Push is probably better. However we don't block writes if the epoch doesn't bump here - I'm not sure if we should do that however.

bharathv replied:
maybe in an extreme case if the epoch doesn't bump for like multiple hours?

rockwotj replied:
Okay I've implemented this in https://github.com/redpanda-data/redpanda/pull/26947/commits/de35cd7bfd87160e333fef9ec3f6393376ba03b0.

It's a little complex, so I'm going to extract this bit into a "two stage cache" abstraction and test that.

bharathv commented on src/v/cluster/cluster_epoch_generator.cc:
```diff
@@ -0,0 +1,130 @@
+              new_epoch,
+              _cached_epoch);
+            _cached_epoch = new_epoch;
+            _cached_epoch_time = ss::lowres_clock::now();
```
if the epoch didn't change (eg: under an error condition), should this be updated?

rockwotj replied:
I think so yes. I think it's OK if there is a longer period where the epoch doesn't change, we'll just have a bigger window of L0 objects in this epoch.

rockwotj replied:
We can tune the refresh interval if this becomes a problem.

bharathv commented on src/v/cluster/cluster_epoch_generator.cc:
```diff
@@ -0,0 +1,130 @@
+    model::offset last_offset = co_await stm.bootstrap_committed_offset();
+    while (!as.abort_requested()) {
+        try {
+            co_await ss::sleep_abortable<ss::lowres_clock>(
```
wonder if there are cases where a parition moved from a node that has seen a higher epoch to a node that is stuck on a lower epoch in this sleep .. there will be a mysterious 1 min window of failed epoch checks.

rockwotj replied:
Certainly this will be possible. I don't know how we can fix that however?

also for leaderless produce, I feel like this will be quite an issue...

mmaslankaprv replied:
partition move is a controller command so that should not be a problem really as for the command to be applied the offset must be committed so one implies the other

rockwotj replied:
It's mostly just leadership changes then?

bharathv replied:
>  I don't know how we can fix that however?

I was thinking there should be feedback from users of this epoch service to notify they have seen a larger epoch somewhere that breaks this sleep.

rockwotj replied:
I added cache invalidation to this service, but it's not yet wired into cloud topics. TBD on that part.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69511-->
<!--3a6a39e5-1da2-44f4-a58c-d0a261ef8727-->
<!--title-->
## Retry command for [Build#69511](https://buildkite.com/redpanda/redpanda/builds/69511)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/scram_test.py::ScramBootstrapUserTest.test_invalid_scram_mechanism@{"expect_fail":true,"mechanism":"sCrAm-ShA-512"}
```
<!--/body-->

mmaslankaprv commented on src/v/cluster/cluster_epoch_generator.cc:
```diff
@@ -0,0 +1,130 @@
+            last_offset = new_offset;
+            continue;
+        }
+        auto result = co_await stm.quorum_write_empty_batch(
```
this should only happen on the controller leader

rockwotj replied:
duplicate of https://github.com/redpanda-data/redpanda/pull/26947#discussion_r2225257843

mmaslankaprv commented on src/v/cluster/cluster_epoch_generator.cc:
```diff
@@ -0,0 +1,130 @@
+
+// NOLINTNEXTLINE(*-avoid-reference-coroutine-parameters)
+ss::future<> force_increment_loop(controller_stm& stm, ss::abort_source& as) {
+    model::offset last_offset = co_await stm.bootstrap_committed_offset();
```
I am wondering why did you decided to use `bootstrap_committed_offset()` instead of just committed offset ?

To be perfectly safe on the leader that the epoch it sees it the most recent once we should call `stm.quorum_write_empty_batch` this way a replication will happen making sure the leader is up to date

mmaslankaprv replied:
so basically before returning any `epoch` and after every leadership transfer just do `quorum_write_empty_batch`

rockwotj replied:
I ran into issues where the controller_stm hasn't yet replayed the log yet and the committed offset is `model::offset::min()`

The inner loop doesn't need the wait call in there, but I figured it wasn't worth it to expose a new method just for this

rockwotj replied:
Still using this initially, it helps solve some races during the bootup sequence and staying out of exponential retries. Let me know if see forsee issues with this.

rockwotj committed:
cluster: update epoch service

Update the loop to only run on the controller leader. This epoch is
periodically "frozen". Other cores will poll the controller leader for
this epoch value.

rockwotj committed:
cluster_epoch_service: use controller RPC to fetch latest epoch

In the case that we aren't the leader, we need to make an RPC to get the
latest epoch from the controller leader.

rockwotj commented:
Okay took a pass at updating the epoch generator based on our discussion from Wednesday.

rockwotj commented:
Force pushes: rebase

rockwotj committed:
cluster/epoch_service: implment two stage caching

stage one: just fire off async updates
stage two: block requests until the updates come

bharathv commented on src/v/cluster/types.h:
```diff
@@ -3661,6 +3661,43 @@ struct update_mirror_topic_state_response
+      serde::compat_version<0>> {
+    using rpc_adl_exempt = std::true_type;
+
+    errc ec{errc::success};
```
nit: may be better to have non success default.

rockwotj replied:
I'n not sure what I would pick, and the normal pattern for these RPCs is to default to success. I'll leave it for now.

bharathv commented on src/v/cluster/cluster_epoch_service.h:
```diff
@@ -0,0 +1,116 @@
+//
+// The cluster epoch is a monotonically increasing value that is currently used
+// in the Cloud Topics's L0 implementation.
+class cluster_epoch_service
```
nit: may be add some detail on how the epochs are propagated to shards (some high level design comment).

bharathv commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -0,0 +1,367 @@
+    }
+
+private:
+    void start_service_loop() {
```
mark noexcept?

rockwotj replied:
done

bharathv commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -0,0 +1,367 @@
+            return ss::now();
+        });
+    }
+    void stop_service_loop() {
```
same

rockwotj replied:
done

bharathv commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -0,0 +1,367 @@
+        co_await _queue.shutdown();
+        // Now that the queue won't accept new tasks, we can safely stop the
+        // service loop if it's running.
+        _abort_source.request_abort();
+        co_await std::exchange(_loop, std::nullopt)
+          .or_else([]() { return std::make_optional(ss::now()); })
+          .value();
```
nit can be factored in a helper (same logic in stop_service_loop)

rockwotj replied:
Different now with the exception handling.

bharathv commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -0,0 +1,367 @@
+            _abort_source.request_abort();
+            return std::exchange(_loop, std::nullopt)
+              .or_else([]() { return std::make_optional(ss::now()); })
+              .value();
```
handle exception and ignore / log?

rockwotj replied:
It **should** never throw, but I will add the logic anyways

rockwotj replied:
done

bharathv commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -0,0 +1,367 @@
+      controller_stm_shard, &cluster_epoch_service::get_current_epoch);
+    if (!maybe_epoch) {
+        try {
+            maybe_epoch = co_await fetch_leader_epoch();
```
wondering if every shard should ddos the leader..  since all the shards are routing to `controller_stm_shard`, it may just fetch once from the leader (if it is not the leader)?

rockwotj replied:
I think there are two approaches:

1. change the fetches on other shards to ask shard0 for it's epoch. Maybe cache invalidation becomes tricky.
2. Shard0 pushes it's epoch to other shards. I have so far refrained from a push model so that if cloud topics isn't in use, nothing happens.

rockwotj replied:
Went with approach #1 for now.

bharathv commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -0,0 +1,367 @@
+    _gate.check();
+    // This is safe to check without the lock, because we only use the lock to
+    // limit the number of cross shard calls and RPCs.
+    if (_cached_epoch == observed_epoch) {
```
shouldn't this be 

if (_cached_epoch < observed_epoch)

rockwotj replied:
I need to clarify which epoch this is. I was thinking you would pass in the local epoch that caused an out of sequence error, **not** whatever the higher epoch that caused the error, because we really don't have a way to plumb this back on the write path (but maybe the STM knows how to invalidate the cache, which seems like a leak of abstraction possibly). I do think only passing around shard local epochs is better, but I need to clarify the docs a bunch.

bharathv replied:
ah ok.. the name `observed_epoch` confused me.

rockwotj replied:
Renamed

bharathv reviewed (commented):
looks great.. couple of comments.

rockwotj commented:
Force pushes: address some review feedback. Not quite done yet, I also need to:

- [x] Add tests (I started this with making Clock a template param now, and abstracting the "go ask the leader for the latest).
- [x] Have shards ask shard0 and only shard0 does leader RPCs (with updates to cache invalidation).
- [x] Deeper explanation of this full component in a header comment.
- [x] Clarify docs for cache invalidation

mmaslankaprv commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -0,0 +1,407 @@
+  typename Clock::duration timeout) {
+    auto cluster_leader = leaders->local().get_leader(model::controller_ntp);
+    if (!cluster_leader) {
+        throw std::runtime_error(
```
i am wondering if it would make sense to to use `result` here for error handling like other cluster services does

rockwotj replied:
I pushed a change for this here: https://github.com/redpanda-data/redpanda/commit/50cb4b1620

Do you think this is better?

rockwotj replied:
Tacked that onto this PR

mmaslankaprv commented on src/v/cluster/cluster_epoch_service.h:
```diff
@@ -0,0 +1,117 @@
+    // Returns the current epoch (with caching) for the cluster.
+    //
+    // May be called on any shard.
+    ss::future<int64_t> get_cached_epoch();
```
consider introducing a `named_type` for the cluster epoch

rockwotj replied:
there is one in the cloud topics subsystem, and I don't want to depend on that because of cycles... Maybe I will move it to model

mmaslankaprv commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -0,0 +1,407 @@
+    // Shutdown this state, stopping the service loop if it's running.
+    //
+    // Must be called before destruction.
+    ss::future<> stop() {
```
i think we need a corresponding start method where we would check if current node is a controller leader to make sure we do not missed the notification

rockwotj replied:
Done

rockwotj committed:
cluster/tests: add epoch_service_test

rockwotj committed:
cluster/epoch_service: only make RPCs from shard0

This way all shards are not spamming another node with epoch fetches.

rockwotj committed:
cluster/epoch_service: support node-wide cache invalidation

If we have proof another node has seen a higher epoch, we can go ahead
and invalidate all of our caches.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69691-->
<!--46f4cb27-389d-49c7-bdb9-3aca956b8887-->
<!--13360845-7feb-4982-b295-28c17f30dfc1-->
<!--614d5ea8-8e93-425c-beea-056db6853676-->
<!--dbac0d0b-c107-4e2d-98dd-fc8d374f23ca-->
<!--title-->
## Retry command for [Build#69691](https://buildkite.com/redpanda/redpanda/builds/69691)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTest.test_write
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTxTest.test_write
```
<!--/body-->

rockwotj committed:
cluster/epoch_service: prevent UB in exception method

bharathv commented on src/v/cluster/cluster_epoch_service.h:
```diff
@@ -0,0 +1,154 @@
+// 1. Shard0 on each broker either fetches the epoch from itself (because it is
+//    controller leader) or it makes an RPC to go and fetch from the leader its
+//    current epoch.
+// 2. None shard0 will go fetch the epoch from shard0. If shard0 needs an update
```
nit: Non

rockwotj replied:
done

bharathv commented on src/v/cluster/cluster_epoch_service.h:
```diff
@@ -0,0 +1,154 @@
+    // epoch that caused the sequence violation (ie. that you got back from
+    // `get_cached_epoch`) in order to actually invalidate the cache.
+    ss::future<>
+    invalidate_epoch_cache(int64_t epoch_causing_sequence_violation);
```
mega nit: monotonicity violation? sequence maybe confusing

rockwotj replied:
done

bharathv commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -0,0 +1,460 @@
+template<typename Clock>
+ss::future<> cluster_epoch_service<Clock>::invalidate_epoch_cache(
+  int64_t epoch_causing_sequence_violation) {
+    _gate.hold();
```
does it work without assigning to a var?

rockwotj replied:
nope, fixed. Good catch

bharathv commented on src/v/cluster/cluster_epoch_service.cc:
```diff
@@ -0,0 +1,460 @@
+    auto maybe_epoch = co_await this->container().invoke_on(
+      controller_stm_shard, &cluster_epoch_service::get_current_epoch);
+    auto update_time = Clock::now();
+    if (!maybe_epoch && ss::this_shard_id() == 0) {
```
nit: == controller_stm_shard

rockwotj replied:
done

bharathv reviewed (commented):
lgtm , only gate assignment to a var is a blocking comment

rockwotj committed:
cluster/epoch_service/tests: prevent flakes in cluster epoch test

The test used to flake because the draining of tasks didn't account for
outstanding cross core tasks.

rockwotj committed:
cluster/epoch_service: use std::expected instead of exceptions

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69772-->
<!--050390ce-418c-41c8-9d20-c110d8407684-->
<!--58164509-e7ba-4187-ab28-f56c6d361afd-->
<!--c39a3317-0fdf-44b2-98bf-d555ce401cda-->
<!--b51ae82c-177d-47d4-a766-a5b8dd912d96-->
<!--6e9faedf-fafa-47dd-a2fa-b6364f802f06-->
<!--61e6d3fa-9ec8-41c3-b9b6-4ec942e3dd1b-->
<!--d7e97bd0-904c-4fd1-ae79-c30efad5ce6a-->
<!--d4fd56a6-cbd1-4c12-b668-a014db945158-->
<!--title-->
## Retry command for [Build#69772](https://buildkite.com/redpanda/redpanda/builds/69772)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTest.test_write
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTxTest.test_write
```
<!--/body-->

rockwotj commented:
Force push: move cloud topics back in the startup sequence, since it depends on the controller being up and RPCs ready.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69790-->
<!--024f2b4c-aa65-4af4-aaf5-e5628e4066fc-->
<!--9013cebb-919c-4457-8270-2e6585aa80a2-->
<!--title-->
## Retry command for [Build#69790](https://buildkite.com/redpanda/redpanda/builds/69790)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
```
<!--/body-->

rockwotj committed:
cluster/epoch_service: add warn logs when failing to update epoch

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69882-->
<!--42f3470d-c656-4f5d-9d47-01481787c449-->
<!--042a0871-f149-4268-b887-09bcb776e271-->
<!--title-->
## Retry command for [Build#69882](https://buildkite.com/redpanda/redpanda/builds/69882)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
```
<!--/body-->

rockwotj committed:
cluster/epoch_service: query raft0 leadership directly

Instead of waiting for metadata to propigate, just query the raft group
directly for leadership.

rockwotj commented:
Force push: rebase & fix conflicts

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69921-->
<!--13a1b300-5052-4832-b252-13b33c3fa5a0-->
<!--1d2653f5-1363-47f5-8309-645bb844a81d-->
<!--8c978f2e-36c9-4271-a0f7-03654e8ace2c-->
<!--e5d5f5be-8cb0-458f-a9ab-bce598cbc3e4-->
<!--title-->
## Retry command for [Build#69921](https://buildkite.com/redpanda/redpanda/builds/69921)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
```
<!--/body-->

rockwotj committed:
cluster/epoch_service: retry until raft0 has a leader

This involves plumbing an abort_source all over, but I think we have to
in order to cleanly shutdown.

dotnwat commented on src/v/cluster/cluster_epoch_generator.cc:
```diff
@@ -0,0 +1,130 @@
+              clusterlog.debug,
+              "Forced increment of cluster epoch to: {}",
+              last_offset);
+        } else if (result.error() != raft::errc::not_leader) {
```
wondering if you also need to handle all these these 🤣 

```
              result.errc == cluster::errc::not_leader_controller
              || result.errc == cluster::errc::no_leader_controller
              || result.errc == raft::errc::not_leader) {
```

rockwotj replied:
This is a raft API, so I only need to handle raft error codes

dotnwat commented on src/v/cluster/cluster_epoch_generator.cc:
```diff
@@ -0,0 +1,130 @@
+            _cached_epoch_time = ss::lowres_clock::now();
+        }
+    }
+    co_return _cached_epoch;
```
so this is to avoid poking core 0 unnecessarily frequently?

rockwotj replied:
yes or making cross core calls

dotnwat commented on src/v/cluster/cluster_epoch_generator.cc:
```diff
@@ -0,0 +1,130 @@
+            last_offset = new_offset;
+            continue;
+        }
+        auto result = co_await stm.quorum_write_empty_batch(
```
would some of these co_await stm's ever throw? might make sense to wrap the entire increment loop in a try block so the loop will auto restart?

dotnwat replied:
this seems to have been completely rewritten in later commits

rockwotj replied:
yes sorry

dotnwat reviewed (commented):
looks pretty good but holding off on review there seems to be a lot of churn. just lemme know when things stabilize.

the one thing i was hoping to get @bharathv @mmaslankaprv take on is if the way the epoch service is built in this PR is going to support expansion of the API to include `get_current_partitions() → (partitions, epoch)` (from the RFC) which is required for garbage collection. I think it will need access to the set of partitions it a way that lets us determine the point-in-time (committed offset) associated with a snapshot of the partition set.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69993-->
<!--90609c07-9454-485a-9a37-32de235ce728-->
<!--title-->
## Retry command for [Build#69993](https://buildkite.com/redpanda/redpanda/builds/69993)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
```
<!--/body-->

mmaslankaprv commented:
> looks pretty good but holding off on review there seems to be a lot of churn. just lemme know when things stabilize.
> 
> the one thing i was hoping to get @bharathv @mmaslankaprv take on is if the way the epoch service is built in this PR is going to support expansion of the API to include `get_current_partitions() → (partitions, epoch)` (from the RFC) which is required for garbage collection. I think it will need access to the set of partitions it a way that lets us determine the point-in-time (committed offset) associated with a snapshot of the partition set.

If i understand that correctly the api `get_current_partitions()` (without any optimization) will have to iterate over all the partitions and query them for the latest epoch. I think this is orthogonal to the epoch source. The question is how to deliver the last seen partition epoch information to all the nodes that needs it.

rockwotj committed:
rptest: deflake cloud topics test

The test was originally written to assume that we wrote
1 L1 object per NTP, but that changed (not the test tho).
However, prior to this PR we used to upload fast enough
that there was at least 5 objects. Now that we have to
wait a bit longer to fetch an epoch from the controller
leader we batched better and got less objects than the
number of partitions :disappointed:

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70020-->
<!--ff87a8b7-00cb-4fad-8014-6dba0b0154ea-->
<!--010ba917-8922-4c58-8b1e-afc9f8a7a2c2-->
<!--fd7039a2-a6ef-4cf9-90ce-8156f4c5cd24-->
<!--88f8c4d9-c34c-431d-bfc8-73c034664c7a-->
<!--title-->
## Retry command for [Build#70020](https://buildkite.com/redpanda/redpanda/builds/70020)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70031-->
<!--18b1d987-1742-4c95-8815-ef23cff60130-->
<!--50c02b7d-140e-4de1-8afe-21493775cf7c-->
<!--e6a4eb4f-4bef-430e-9cce-80864cde3c9c-->
<!--1467ed46-e3ed-4bc2-aab2-f02366657823-->
<!--title-->
## Retry command for [Build#70031](https://buildkite.com/redpanda/redpanda/builds/70031)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
```
<!--/body-->

rockwotj committed:
cloud_topics: use watchdog for shutdown

rockwotj committed:
ct/reconciler: log object paths

The ducktape test is failing here and I think it's because we're
batching more things together. Where before we assumed maybe that each
NTP was uploaded seperately?

rockwotj commented:
Sorry for everyone getting spam following along...

I did finally find the issue. The test was originally written to assume that we wrote 1 L1 object per NTP, but that changed (not the test tho). However, prior to this PR we used to upload fast enough that there was at least 5 objects. Now that we have to wait a bit longer to fetch an epoch from the controller leader we batched better and got less objects than the number of partitions :disappointed:

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70050-->
<!--faf188ef-18a1-4c12-a271-85b9dd1db50b-->
<!--2fe5dad6-1ebc-4d5e-be89-4faeb923b826-->
<!--c0376fae-98ab-40bd-a66b-2aae24adbd04-->
<!--title-->
## Retry command for [Build#70050](https://buildkite.com/redpanda/redpanda/builds/70050)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics_test.py::CloudTopicsTest.test_reconciler_uploads@{"cloud_storage_type":2}
```
<!--/body-->

rockwotj committed:
cloud_topics/batcher: lazily fetch epoch

Eagerly fetching the epoch on startup has some annoying side effects of
temporary errors. So just handle those temporary errors and error the
writes. In practice the error rejecting writes due to no epoch should
only happen if the cluster is unhealthy and the controller is unable
to elect a leader.

dotnwat commented on src/v/cloud_topics/level_zero/batcher/batcher.cc:
```diff
@@ -164,7 +162,23 @@ ss::future<result<bool>> batcher<Clock>::run_once() noexcept {
         }
 
-        aggregator<Clock> aggregator{object_id::create(epoch)};
+        auto epoch_fut = co_await ss::coroutine::as_future<cluster_epoch>(
+          _cluster_services->current_epoch(&_as));
```
aggressively 😎

dotnwat reviewed (commented):
looks pretty solid to me. thanks!
